### PR TITLE
Add `stderr` support

### LINKED ISSUE #301: Auto-detect support for `stderr`
[Issuehunt badges](https://issuehunt.io/r/chalk/chalk/issues/301)

I want use colors only on `console.error` calls which is typically terminal.
I suggest some API extension, very easy idea is chaining like this:
```js
const coloredOutput = chalk.stderr.red('RED ALERT');
```
This can be used more easily like this:
```js
const chalk = require('chalk').stderr;
const coloredOutput = chalk.red('RED ALERT');
```

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $80.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
### Submitted pull Requests
- [#359 Add chalk.stderr and chalk.stderr.supportsColor](https://issuehunt.io/r/chalk/chalk/pull/359)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/chalk/chalk/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### LINKED ISSUE #318: Invalid type definition for `supports-color` reexport

### PR DESCRIPTION
**Pull request by langpavel**
This add support for `require('chalk/stderr')`

Fixes #301 
Fixes #318

- [x] tests
- [x] documentation (may need rewording)
- [x] typescript definition (may be incorrect), see comment
- [x] flow definition

### CODE REVIEW

File: stderr.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+expectType<string>(chalk`Hello {bold.red ${name}}`);
+
+// -- Color methods --
+expectType<colorReturn>(chalk.hex('#DEADED'));
+
+// -- Complex --
+expectType<string>(chalk.red.bgGreen.underline('foo'));
+expectType<string>(chalk.underline.red.bgGreen('foo'));
+
+// -- No more needed as tests are same as for stdout
```
[the remark this answers is not in the document]
Comment (Qix-) [reply stderr.test-d.ts]: Further, comments that explain why tests don't exist aren't very helpful. You can just remove that last comment about "no more needed".

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+	"editor.formatOnSave": false
+}
```
[the remark this answers is not in the document]
Comment (langpavel) [reply .vscode/settings.json]: Oh, sorry

**Comment by langpavel** [comment]
I really don't know how to (where I should) add new type definitions for `chalk/stderr`
Now I have separate file for typescript but it seems like tsd-check is ignoring this file
Similar — flow definitions. Really don't know where I should add.
From type point of view, `chalk/stderr` is only reexport of `chalk`
Can anyone help me/point me at good TS/flow resource where I can learn?

**Comment by langpavel** [comment]
@sindresorhus @Qix- This is everything from me.
If you can review again, please :pray: 
Thanks

File: readme.md [hunk 3]
```diff
@@ -55,9 +55,13 @@ $ npm install chalk
 ## Usage
 
 ```js
+// for stdout: console.log, console.info, console.debug
 const chalk = require('chalk');
+// for stderr: console.error, console.warn
+const cherr = require('chalk/stderr');
 
 console.log(chalk.blue('Hello world!'));
+console.error(cherr.red('Oh, colored error!'));
```
Comment (sindresorhus) [inline readme.md]: I don't want `stderr` examples here. It's not something most would need. You can add one example to the `chalk/stderr` section instead.

File: stderr.d.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import Chalk from '.';
+
+declare const chalk: Chalk;
+
+export default chalk;
```
Comment (sindresorhus) [inline stderr.d.ts]: Does `export default chalk from '.';` not work?

File: test/_fixture.js [hunk 5]
```diff
@@ -1,4 +1,6 @@
 'use strict';
 const chalk = require('..');
+const cherr = require('../stderr');
```
Comment (sindresorhus) [inline test/_fixture.js]: `cherr` => `chalkStderr`

File: stderr.test-d.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import {expectType} from 'tsd-check';
+import cherr, {Level, Chalk, ColorSupport} from './stderr';
```
Comment (sindresorhus) [inline stderr.test-d.ts]: `cherr` => `chalkStderr`

File: test/_flow.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
 // @flow
 import chalk from '..';
+import cherr from '../stderr';
```
Comment (sindresorhus) [inline test/_flow.js]: `cherr` => `chalkStderr`

File: test/stderr.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import path from 'path';
+import test from 'ava';
+import execa from 'execa';
+
+// Spoof supports-color
+require('./_supports-color')(__dirname);
+
+const cherr = require('../stderr');
```
Comment (sindresorhus) [inline test/stderr.js]: `cherr` => `chalkStderr`

File: readme.md (hunk 3 above)
Comment (langpavel) [reply readme.md]: IMHO this is place where requirement for separate this two is most visible.
At most I can split this to two snippets or add notice under original

File: stderr.d.ts (hunk 4 above)
Comment (langpavel) [reply stderr.d.ts]: Oh, I really don't know if that construct is valid

File: stderr.test-d.ts (hunk 6 above)
Comment (langpavel) [reply stderr.test-d.ts]: Agree

File: readme.md (hunk 3 above)
Comment (langpavel) [reply readme.md]: Reverted to previous with added notice

File: stderr.d.ts (hunk 4 above)
Comment (langpavel) [reply stderr.d.ts]: No, your suggestion does not work.
If you wanna play with it you can :smiling_imp:

**Comment by sindresorhus** [comment]
Hmm, seeing how this actually looks in practice, I'm starting to wonder whether we made the wrong call in wanting `chalk/stderr`. The ergonomics seem very verbose if you need to output to both stdout and stderr...

```js
const chalk = require('chalk');
const chalkStderr = require('chalk/stderr');

...

chalk.red('foo');

...

chalkStderr.blue('bar');
```

vs.


```js
const chalk = require('chalk');

...

chalk.red('foo');

...

chalk.stderr.blue('bar');
```

Maybe I'm overthinking it. Just thought I would mention it before we're locked into `chalk/stderr` for eternity.

**Comment by langpavel** [comment]
As you said, it's uncommon to use both outputs simultaneously.
I like extra import instead of repeating `chalk.stderr`. Even if you bind it to global const, you will not save nothing.

**Comment by langpavel** [comment]
@Qix- what do you think?

**Comment by Qix-** [comment]
Someone else suggested `require('chalk').stderr`.

**Comment by sindresorhus** [comment]
I totally forgot about this...

I have been thinking a lot about it and I still think

```js
const chalk = require('chalk');
chalk.red('foo');
chalk.stderr.blue('bar');
```

makes the most sense.

@langpavel Would you be able to update the PR?

**Comment by Qix-** [comment]
@sindresorhus FWIW that way still allows for `const chalk = require('chalk').stderr`, so SGTM.

**Comment by sindresorhus** [comment]
@Qix- Yes, however, what I don't want is `import {stderr} from 'chalk';`. `.stderr` should be attached to the default export in TS/Flow: https://github.com/chalk/chalk/blob/2ca015c4c537896dd378b94ecbbdec935a7b1b52/index.d.ts#L274

**Comment by langpavel** [comment]
I'm really sorry but haven't time. Can anybody take this? :pray:

**Comment by sindresorhus** [comment]
@langpavel Doesn't look like anyone offered to take this on. I'm gonna close this until you or someone else has time to finish this ;)

**Pull request closed without merging**